File: src/lib/undoJournal.ts
import type Database from "@tauri-apps/plugin-sql";

/**
 * The undo journal: every change to the lab record writes its own inverse, as SQL,
 * inside the same statement that makes the change.
 *
 * Undo used to copy the whole database file before every save, across the IPC
 * boundary as a JSON integer array: seconds on a lab-sized database, and every
 * undo race lived in that window. Here nothing is copied. A trigger on each
 * journaled table appends the statement that would reverse the row it just
 * touched, so the record of how to go back is written atomically with the change
 * and costs the size of the change, not the size of the database.
 *
 * An undo entry is a RANGE of the journal, the head either side of its action.
 * Replaying it newest first returns every journaled table to exactly its state at
 * the start of the range. The replay is itself journaled, so the rows it writes
 * are the redo. The replay runs in one transaction in Rust
 * (src-tauri/src/undo_journal.rs), because statements sent through
 * tauri-plugin-sql's pool cannot share one.
 *
 * Each inverse CARRIES ITS OWN PRECONDITION: it matches the row only while the row
 * still holds exactly what the change left there. A replay is one entry's range of
 * the journal, not everything after it, so a write that landed outside that range
 * -- the sync timer draining a viewer's request between an Undo and the Redo of
 * it -- would otherwise be half-erased by a full-row restore. Guarded, that
 * statement matches no row, and the replay refuses with nothing changed
 * (`undo_journal.rs` requires every statement to touch exactly one row).
 *
 * The triggers are persistent, not TEMP: the pool has several connections, and a
 * TEMP trigger exists on one connection only. They are created at runtime by
 * getDb(), with no numbered migration, like `samples.embedding_notes` (AGENTS.md):
 * a migration records its version in the file, and the build in use refuses a
 * database recording a version it does not know, so rolling back would break.
 */

/**
 * What a replay is refused with when a row it would put back is no longer as the
 * action left it, so restoring the whole row would erase whatever changed it.
 * The wording is the command's contract: `CHANGED_SINCE` in
 * src-tauri/src/undo_journal.rs, which is what the running app actually raises,
 * and src/test/undoJournalCommands.ts models. Change all three together.
 */
export const CHANGED_SINCE =
  "Cannot undo or redo that step: the records it would put back have changed since. Nothing was changed.";

/** Session and bookkeeping tables. Undo has never rewound these (restoreDbPreservingSession). */
const NOT_JOURNALED = new Set([
  "users",
  "app_settings",
  "schema_meta",
  "undo_journal",
  "_sqlx_migrations",
]);

const TRIGGER_PREFIX = "undo_journal_";

const ident = (name: string) => `"${name.replace(/"/g, '""')}"`;
/** A SQL string literal holding `text`, for splicing fixed text into a trigger body. */
const lit = (text: string) => `'${text.replace(/'/g, "''")}'`;

export const JOURNAL_TABLE_SQL =
  `CREATE TABLE IF NOT EXISTS undo_journal (` +
  `seq INTEGER PRIMARY KEY AUTOINCREMENT, stmt TEXT NOT NULL, at TEXT NOT NULL DEFAULT (datetime('now')))`;

/**
 * How much journal is worth carrying.
 *
 * The cost is BYTES, not rows: the journal rides inside the database file, so it
 * is copied into every backup and uploaded whole in every snapshot the
 * workstation publishes. Rows vary by two orders of magnitude - an UPDATE inverse
 * writes each of a table's columns twice, once to set it back and once to guard
 * it, and a block's four note fields hold whatever was typed at the bench - so a
 * row count says little about the payload. Five megabytes is the ceiling: a
 * fraction of a lab-sized database (about 23 MB), and far more journal than the
 * hundred-step stack can reach in ordinary use, where `pruneJournal` keeps it
 * near the live stack anyway. The row count and the age are secondary guards,
 * for a journal that is small in bytes and long in rows or in days.
 */
const KEEP_BYTES = 5 * 1024 * 1024;
const KEEP_ROWS = 20_000;
const KEEP_DAYS = 14;

/**
 * Forget journal rows past the bound above, oldest first.
 *
 * Run at open (db.ts), where nothing is live yet: the undo stack is empty until
 * the history is hydrated, and that hydration is anchored to the journal's ends,
 * so a history needing a row this trimmed is dropped rather than half replayed
 * (undoPersist.ts). A replay is never left with part of its range.
 */
export function journalTrimStatements(): string[] {
  const running = `SUM(LENGTH(CAST(stmt AS BLOB))) OVER (ORDER BY seq DESC)`;
  return [
    `DELETE FROM undo_journal WHERE at < datetime('now', '-${KEEP_DAYS} days')`,
    `DELETE FROM undo_journal WHERE seq NOT IN (` +
      `SELECT seq FROM (SELECT seq, ${running} AS bytes FROM undo_journal) WHERE bytes <= ${KEEP_BYTES})`,
    `DELETE FROM undo_journal WHERE seq NOT IN (SELECT seq FROM undo_journal ORDER BY seq DESC LIMIT ${KEEP_ROWS})`,
  ];
}

/**
 * The three triggers that journal `table`, whose columns are `columns`.
 *
 * `rowidAlias` says a column already IS the rowid (an INTEGER PRIMARY KEY), which
 * then carries the row's identity; otherwise the rowid is written out beside the
 * columns, so a re-inserted row comes back under the id everything else refers to.
 *
 * The inverse of an INSERT and of an UPDATE is guarded by the row's post-change
 * values, so it applies only while nothing else has touched that row since; the
 * inverse of a DELETE needs no guard, because it re-inserts under the same rowid
 * and SQLite refuses that outright if anything has taken the row's place.
 */
export function journalTriggers(
  table: string,
  columns: string[],
  rowidAlias: boolean,
): Array<{ name: string; sql: string }> {
  const t = ident(table);
  const name = (op: string) => `${TRIGGER_PREFIX}${table}_${op}`;
  const values = (row: "old") =>
    columns.map((c) => `quote(${row}.${ident(c)})`).join(` || ',' || `);
  const assignments = columns
    .map(
      (c, i) =>
        `${lit(`${i === 0 ? "" : ","}${ident(c)}=`)} || quote(old.${ident(c)})`,
    )
    .join(" || ");
  const untouched = columns
    .map((c) => `${lit(` AND ${ident(c)} IS `)} || quote(new.${ident(c)})`)
    .join(" || ");
  const colList = (rowidAlias ? columns : ["rowid", ...columns])
    .map((c) => (c === "rowid" ? c : ident(c)))
    .join(",");
  const rowValues = rowidAlias
    ? values("old")
    : `old.rowid || ',' || ${values("old")}`;
  return [
    {
      name: name("insert"),
      sql: `CREATE TRIGGER ${ident(name("insert"))} AFTER INSERT ON ${t} BEGIN
  INSERT INTO undo_journal(stmt) VALUES (${lit(`DELETE FROM ${t} WHERE rowid=`)} || new.rowid || ${untouched});
END`,
    },
    {
      name: name("update"),
      sql: `CREATE TRIGGER ${ident(name("update"))} AFTER UPDATE ON ${t} BEGIN
  INSERT INTO undo_journal(stmt) VALUES (${lit(`UPDATE ${t} SET `)} || ${assignments} || ' WHERE rowid=' || new.rowid || ${untouched});
END`,
    },
    {
      name: name("delete"),
      sql: `CREATE TRIGGER ${ident(name("delete"))} AFTER DELETE ON ${t} BEGIN
  INSERT INTO undo_journal(stmt) VALUES (${lit(`INSERT INTO ${t}(${colList}) VALUES(`)} || ${rowValues} || ')');
END`,
    },
  ];
}

/**
 * The statements that bring the journal and its triggers into line with the
 * tables this file has now, or none when they already are.
 *
 * Worked out on every open, after the runtime schema has converged, so a column
 * added by a migration or by ensureRuntimeSchema is journaled from its first
 * write. A trigger whose text already matches is left alone, so an unchanged
 * schema costs a few reads and no write. Only reads here: the statements are run
 * together, in one transaction, by the `undo_journal_install` command, so a
 * half-installed set of triggers cannot exist.
 */
export async function journalInstallStatements(db: Pick<Database, "select">): Promise<string[]> {
  const statements: string[] = [];
  const journal = await db.select<Array<{ name: string }>>(`PRAGMA table_info(undo_journal)`);
  if (journal.length === 0) statements.push(JOURNAL_TABLE_SQL);
  const tables = await db.select<Array<{ name: string }>>(
    `SELECT name FROM sqlite_master WHERE type = 'table' AND name NOT LIKE 'sqlite_%' ORDER BY name`,
  );
  const existing = new Map(
    (
      await db.select<Array<{ name: string; sql: string }>>(
        `SELECT name, sql FROM sqlite_master WHERE type = 'trigger' AND name LIKE '${TRIGGER_PREFIX}%'`,
      )
    ).map((r) => [r.name, r.sql]),
  );
  const wanted = new Set<string>();
  for (const { name: table } of tables) {
    if (NOT_JOURNALED.has(table)) continue;
    const info = await db.select<Array<{ name: string; type: string; pk: number }>>(`PRAGMA table_info(${ident(table)})`);
    const keys = info.filter((c) => c.pk > 0);
    const rowidAlias = keys.length === 1 && keys[0].type.toUpperCase() === "INTEGER";
    for (const trigger of journalTriggers(table, info.map((c) => c.name), rowidAlias)) {
      wanted.add(trigger.name);
      if (existing.get(trigger.name) === trigger.sql) continue;
      statements.push(`DROP TRIGGER IF EXISTS ${ident(trigger.name)}`, trigger.sql);
    }
  }
  for (const name of existing.keys()) {
    if (!wanted.has(name)) statements.push(`DROP TRIGGER IF EXISTS ${ident(name)}`);
  }
  return statements;
}

File: src-tauri/src/undo_journal.rs
//! Undo and redo as one transaction each, on a connection of their own.
//!
//! Every change to the lab record writes its own inverse, as SQL, into
//! `undo_journal` through triggers the frontend installs (`src/lib/undoJournal.ts`).
//! An undo entry is a range of that journal: the rows its action wrote. Undo
//! replays the range, newest first, and the replay's own rows are the range
//! that redoes it. Each inverse carries its own precondition, so it matches its
//! row only while that row still holds what the change left there; a statement
//! that matches no row means something outside the range has touched it since,
//! and the whole replay is refused rather than half applied.
//! A replay is many statements, and the frontend's writes go through
//! tauri-plugin-sql's connection POOL, where a BEGIN in one call and a COMMIT in
//! the next can land on different connections. So the two things that must be
//! all-or-nothing run here instead, each inside one `BEGIN IMMEDIATE` on a
//! connection opened for it:
//!
//!  * [`revert`]: replay one range of the journal, newest first, then sweep the
//!    audit rows the replay itself provoked, and report the range the replay
//!    wrote (the entry that reverses it);
//!  * [`execute_batch`]: run a list of statements together, used to install or
//!    refresh the journal's triggers so a half-installed set can never exist.
//!
//! On any failure the transaction is rolled back and nothing has changed. The
//! test harnesses model both commands in `src/test/undoJournalCommands.ts`; change
//! the two together.

use std::path::Path;

use sqlx::sqlite::SqliteConnectOptions;
use sqlx::{Connection, SqliteConnection};

/// Open the live database the way a launch does (sqlx's default options: WAL,
/// foreign keys on, a busy timeout), refusing to create a file that is not there.
async fn open(path: &Path) -> Result<SqliteConnection, String> {
    let options = SqliteConnectOptions::new()
        .filename(path)
        .create_if_missing(false);
    SqliteConnection::connect_with(&options)
        .await
        .map_err(|e| format!("Could not open the database: {e}"))
}

/// Take the write lock up front: IMMEDIATE, so a replay that starts by reading
/// the journal cannot be refused the lock halfway through.
async fn begin(conn: &mut SqliteConnection) -> Result<(), String> {
    sqlx::raw_sql("BEGIN IMMEDIATE")
        .execute(&mut *conn)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

/// Commit what `work` did if it succeeded, roll all of it back if it failed, and
/// close the connection either way.
async fn finish<T>(mut conn: SqliteConnection, work: Result<T, String>) -> Result<T, String> {
    let result = match work {
        Ok(value) => sqlx::raw_sql("COMMIT")
            .execute(&mut conn)
            .await
            .map(|_| value)
            .map_err(|e| e.to_string()),
        Err(err) => {
            let _ = sqlx::raw_sql("ROLLBACK").execute(&mut conn).await;
            Err(err)
        }
    };
    let _ = conn.close().await;
    result
}

/// What a replay refuses with when a row it would put back is no longer as the
/// action left it, so restoring the whole row would erase whatever changed it.
pub const CHANGED_SINCE: &str =
    "Cannot undo or redo that step: the records it would put back have changed since. Nothing was changed.";

/// The rows a replay wrote: `(from, to]` of the journal, the range that reverses it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
pub struct Replayed {
    pub from: i64,
    pub to: i64,
}

/// Replay the journal rows in `(from, to]`, newest first (`to` absent: up to the
/// head), so every journaled table is back as it stood at `from`, and return the
/// range of rows the replay itself wrote.
///
/// Only that range, never everything after `from`: an undo replaying everything
/// after its mark would also replay every earlier undo's own rows. Each pair
/// cancels, so the result would be right, but every replay would write back all
/// the ones before it and the journal would double with every undo in a row.
///
/// The replay fires the app's own audit triggers (0010 onward) as it goes. Those
/// rows narrate the replay, not an action, and undo has never written them, so
/// every audit row above `audit_events`' AUTOINCREMENT high-water mark from
/// before the replay is swept out. The high-water mark, not `MAX(id)`: after an
/// undo, the rows a redo puts back sit below it. The sweep is journaled too, so
/// a redo restores the undone action's own audit rows exactly.
///
/// Every journal statement must touch EXACTLY ONE row. Each is an inverse of one
/// row's change, guarded by what that change left there, so none touching a row
/// means something outside the range being replayed has changed it since -- the
/// sync timer draining a viewer's request between an Undo and its Redo, say --
/// and restoring the whole row would silently erase that. Refused with
/// [`CHANGED_SINCE`], rolled back, nothing changed.
pub async fn revert(path: &Path, from: i64, to: Option<i64>) -> Result<Replayed, String> {
    let mut conn = open(path).await?;
    begin(&mut conn).await?;
    let work = replay(&mut conn, from, to).await;
    finish(conn, work).await
}

async fn head(conn: &mut SqliteConnection) -> Result<i64, sqlx::Error> {
    sqlx::query_scalar("SELECT COALESCE((SELECT seq FROM sqlite_sequence WHERE name = 'undo_journal'), 0)")
        .fetch_one(&mut *conn)
        .await
}

async fn replay(conn: &mut SqliteConnection, from: i64, to: Option<i64>) -> Result<Replayed, String> {
    let start = head(conn).await.map_err(|e| e.to_string())?;
    let audit: i64 = sqlx::query_scalar(
        "SELECT COALESCE((SELECT seq FROM sqlite_sequence WHERE name = 'audit_events'), 0)",
    )
    .fetch_one(&mut *conn)
    .await
    .map_err(|e| e.to_string())?;
    let statements: Vec<String> = sqlx::query_scalar(
        "SELECT stmt FROM undo_journal WHERE seq > ? AND seq <= ? ORDER BY seq DESC",
    )
    .bind(from)
    .bind(to.unwrap_or(start))
    .fetch_all(&mut *conn)
    .await
    .map_err(|e| e.to_string())?;
    for statement in &statements {
        let done = sqlx::raw_sql(statement)
            .execute(&mut *conn)
            .await
            .map_err(|e| e.to_string())?;
        if done.rows_affected() != 1 {
            return Err(CHANGED_SINCE.to_string());
        }
    }
    sqlx::query("DELETE FROM audit_events WHERE id > ?")
        .bind(audit)
        .execute(&mut *conn)
        .await
        .map_err(|e| e.to_string())?;
    Ok(Replayed { from: start, to: head(conn).await.map_err(|e| e.to_string())? })
}

/// Run `statements` in order as one transaction: all of them, or none.
pub async fn execute_batch(path: &Path, statements: &[String]) -> Result<(), String> {
    let mut conn = open(path).await?;
    begin(&mut conn).await?;
    let mut work = Ok(());
    for statement in statements {
        if let Err(err) = sqlx::raw_sql(statement).execute(&mut conn).await {
            work = Err(err.to_string());
            break;
        }
    }
    finish(conn, work).await
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;
    use std::sync::atomic::{AtomicUsize, Ordering};

    static SEQ: AtomicUsize = AtomicUsize::new(0);

    fn block_on<F: std::future::Future>(f: F) -> F::Output {
        tauri::async_runtime::block_on(f)
    }

    /// A scratch directory holding one database file, removed when dropped.
    struct Scratch(PathBuf);

    impl Scratch {
        fn new() -> Self {
            let dir = std::env::temp_dir().join(format!(
                "histometer-undo-journal-test-{}-{}",
                std::process::id(),
                SEQ.fetch_add(1, Ordering::SeqCst)
            ));
            let _ = std::fs::remove_dir_all(&dir);
            std::fs::create_dir_all(&dir).unwrap();
            Scratch(dir)
        }
        fn db(&self) -> PathBuf {
            self.0.join("histometer.db")
        }
    }

    impl Drop for Scratch {
        fn drop(&mut self) {
            let _ = std::fs::remove_dir_all(&self.0);
        }
    }

    /// A small lab: one table journaled the way the frontend journals every
    /// table -- every inverse guarded by what the change left in the row, as
    /// `journalTriggers` writes them -- and an audit trigger like the app's own.
    const SCHEMA: &str = "
        CREATE TABLE audit_events (id INTEGER PRIMARY KEY AUTOINCREMENT, summary TEXT NOT NULL);
        CREATE TABLE undo_journal (seq INTEGER PRIMARY KEY AUTOINCREMENT, stmt TEXT NOT NULL);
        CREATE TABLE samples (id INTEGER PRIMARY KEY, note TEXT NOT NULL DEFAULT '');
        CREATE TRIGGER audit_samples AFTER UPDATE ON samples BEGIN
          INSERT INTO audit_events(summary) VALUES ('sample ' || new.id || ' now ' || new.note);
        END;
        CREATE TRIGGER undo_journal_samples_insert AFTER INSERT ON samples BEGIN
          INSERT INTO undo_journal(stmt) VALUES ('DELETE FROM samples WHERE rowid=' || new.rowid || ' AND id IS ' || quote(new.id) || ' AND note IS ' || quote(new.note));
        END;
        CREATE TRIGGER undo_journal_samples_update AFTER UPDATE ON samples BEGIN
          INSERT INTO undo_journal(stmt) VALUES ('UPDATE samples SET id=' || quote(old.id) || ',note=' || quote(old.note) || ' WHERE rowid=' || new.rowid || ' AND id IS ' || quote(new.id) || ' AND note IS ' || quote(new.note));
        END;
        CREATE TRIGGER undo_journal_samples_delete AFTER DELETE ON samples BEGIN
          INSERT INTO undo_journal(stmt) VALUES ('INSERT INTO samples(id,note) VALUES(' || quote(old.id) || ',' || quote(old.note) || ')');
        END;
        CREATE TRIGGER undo_journal_audit_events_insert AFTER INSERT ON audit_events BEGIN
          INSERT INTO undo_journal(stmt) VALUES ('DELETE FROM audit_events WHERE rowid=' || new.rowid || ' AND id IS ' || quote(new.id) || ' AND summary IS ' || quote(new.summary));
        END;
        CREATE TRIGGER undo_journal_audit_events_delete AFTER DELETE ON audit_events BEGIN
          INSERT INTO undo_journal(stmt) VALUES ('INSERT INTO audit_events(id,summary) VALUES(' || quote(old.id) || ',' || quote(old.summary) || ')');
        END;
    ";

    async fn lab(scratch: &Scratch) -> SqliteConnection {
        let options = SqliteConnectOptions::new()
            .filename(scratch.db())
            .create_if_missing(true);
        let mut conn = SqliteConnection::connect_with(&options).await.unwrap();
        sqlx::raw_sql(SCHEMA).execute(&mut conn).await.unwrap();
        conn
    }

    async fn run(conn: &mut SqliteConnection, sql: &str) {
        sqlx::raw_sql(sql).execute(&mut *conn).await.unwrap();
    }

    async fn head(conn: &mut SqliteConnection) -> i64 {
        super::head(conn).await.unwrap()
    }

    async fn rows(conn: &mut SqliteConnection, sql: &str) -> Vec<String> {
        sqlx::query_scalar(sql).fetch_all(&mut *conn).await.unwrap()
    }

    const SAMPLES: &str = "SELECT id || ':' || note FROM samples ORDER BY id";
    const AUDIT: &str = "SELECT id || ':' || summary FROM audit_events ORDER BY id";

    #[test]
    fn a_replay_puts_every_row_back_and_a_second_replay_redoes_it() {
        block_on(async {
            let scratch = Scratch::new();
            let mut conn = lab(&scratch).await;
            run(&mut conn, "INSERT INTO samples(id, note) VALUES (1, 'a'), (2, 'b')").await;
            let mark = head(&mut conn).await;
            let (samples, audit) = (rows(&mut conn, SAMPLES).await, rows(&mut conn, AUDIT).await);

            run(&mut conn, "UPDATE samples SET note = 'c' WHERE id = 1; DELETE FROM samples WHERE id = 2; INSERT INTO samples(id, note) VALUES (3, 'd')").await;
            let (after_samples, after_audit) = (rows(&mut conn, SAMPLES).await, rows(&mut conn, AUDIT).await);

            let redo = revert(&scratch.db(), mark, None).await.expect("the undo replays");
            assert_eq!(rows(&mut conn, SAMPLES).await, samples);
            assert_eq!(rows(&mut conn, AUDIT).await, audit, "the replay's own audit rows are swept");

            revert(&scratch.db(), redo.from, Some(redo.to)).await.expect("the redo replays");
            assert_eq!(rows(&mut conn, SAMPLES).await, after_samples);
            assert_eq!(rows(&mut conn, AUDIT).await, after_audit, "the undone action's audit rows come back");
        });
    }

    #[test]
    fn undoing_every_action_in_a_row_then_redoing_them_grows_the_journal_linearly() {
        block_on(async {
            let scratch = Scratch::new();
            let mut conn = lab(&scratch).await;
            run(&mut conn, "INSERT INTO samples(id, note) VALUES (1, 'n0')").await;
            let start = rows(&mut conn, SAMPLES).await;
            // Twelve actions, each an undo entry: its mark, and where it ends once
            // the next one starts (the undo stack's bookkeeping, src/lib/undo.ts).
            let mut entries: Vec<(i64, Option<i64>)> = Vec::new();
            for n in 1..=12 {
                let mark = head(&mut conn).await;
                if let Some(last) = entries.last_mut() {
                    last.1.get_or_insert(mark);
                }
                entries.push((mark, None));
                run(&mut conn, &format!("UPDATE samples SET note = 'n{n}' WHERE id = 1")).await;
            }
            let end = rows(&mut conn, SAMPLES).await;
            let written = head(&mut conn).await;

            // Each action wrote 3 rows (its change, its audit row, that row's
            // inverse). Each undo and each redo replays one action's worth, so every
            // replay adds a few rows, not a doubling. Checked after every replay, so
            // a doubling journal fails in a handful of steps.
            let mut replays = 0;
            let mut redos = Vec::new();
            while let Some((from, to)) = entries.pop() {
                redos.push(revert(&scratch.db(), from, to).await.expect("an undo replays"));
                replays += 1;
                let grown = head(&mut conn).await - written;
                assert!(grown <= replays * 3 * 4, "{replays} replays grew the journal by {grown} rows");
            }
            assert_eq!(rows(&mut conn, SAMPLES).await, start, "every action is undone");
            while let Some(r) = redos.pop() {
                revert(&scratch.db(), r.from, Some(r.to)).await.expect("a redo replays");
                replays += 1;
                let grown = head(&mut conn).await - written;
                assert!(grown <= replays * 3 * 4, "{replays} replays grew the journal by {grown} rows");
            }
            assert_eq!(rows(&mut conn, SAMPLES).await, end, "every action is redone");
        });
    }

    #[test]
    fn a_replay_that_fails_part_way_changes_nothing() {
        block_on(async {
            let scratch = Scratch::new();
            let mut conn = lab(&scratch).await;
            run(&mut conn, "INSERT INTO samples(id, note) VALUES (1, 'a')").await;
            let mark = head(&mut conn).await;
            run(&mut conn, "UPDATE samples SET note = 'b' WHERE id = 1").await;
            // The OLDEST row after the mark cannot run, so the replay, newest first,
            // applies the rows after it and only then meets it.
            run(&mut conn, &format!("UPDATE undo_journal SET stmt = 'UPDATE no_such_table SET x = 1' WHERE seq = {}", mark + 1)).await;
            assert!(head(&mut conn).await > mark + 1, "a good row is replayed before the bad one");
            let (samples, audit, journal) = (
                rows(&mut conn, SAMPLES).await,
                rows(&mut conn, AUDIT).await,
                rows(&mut conn, "SELECT stmt FROM undo_journal ORDER BY seq").await,
            );

            let err = revert(&scratch.db(), mark, None).await.expect_err("the replay is refused");
            assert!(err.contains("no_such_table"), "{err}");
            assert_eq!(rows(&mut conn, SAMPLES).await, samples, "no row was reverted");
            assert_eq!(rows(&mut conn, AUDIT).await, audit, "no audit row was swept");
            assert_eq!(rows(&mut conn, "SELECT stmt FROM undo_journal ORDER BY seq").await, journal);
        });
    }

    /// The sync timer drains a viewer's request between an Undo and the Redo of it,
    /// touching a row the action touched. The redo would restore that whole row,
    /// erasing the drain's change to it while the rest of the drain's work stayed.
    #[test]
    fn a_replay_is_refused_when_a_row_it_would_put_back_changed_since() {
        block_on(async {
            let scratch = Scratch::new();
            let mut conn = lab(&scratch).await;
            run(&mut conn, "INSERT INTO samples(id, note) VALUES (1, 'a')").await;
            let mark = head(&mut conn).await;
            run(&mut conn, "UPDATE samples SET note = 'b' WHERE id = 1").await;

            let redo = revert(&scratch.db(), mark, None).await.expect("the undo replays");
            assert_eq!(rows(&mut conn, SAMPLES).await, vec!["1:a"]);

            // The drain, outside the range the redo would replay.
            run(&mut conn, "UPDATE samples SET note = 'drained' WHERE id = 1").await;
            let (samples, audit, journal) = (
                rows(&mut conn, SAMPLES).await,
                rows(&mut conn, AUDIT).await,
                rows(&mut conn, "SELECT stmt FROM undo_journal ORDER BY seq").await,
            );

            let err = revert(&scratch.db(), redo.from, Some(redo.to))
                .await
                .expect_err("the redo is refused");
            assert_eq!(err, CHANGED_SINCE);
            assert_eq!(rows(&mut conn, SAMPLES).await, samples, "the drain's write is untouched");
            assert_eq!(rows(&mut conn, AUDIT).await, audit);
            assert_eq!(rows(&mut conn, "SELECT stmt FROM undo_journal ORDER BY seq").await, journal);
        });
    }

    #[test]
    fn a_batch_that_fails_part_way_changes_nothing() {
        block_on(async {
            let scratch = Scratch::new();
            let mut conn = lab(&scratch).await;
            let err = execute_batch(
                &scratch.db(),
                &[
                    "CREATE TABLE installed (x INTEGER)".to_string(),
                    "CREATE TRIGGER broken AFTER INSERT ON no_such_table BEGIN SELECT 1; END".to_string(),
                ],
            )
            .await
            .expect_err("the batch is refused");
            assert!(err.contains("no_such_table"), "{err}");
            let tables = rows(&mut conn, "SELECT name FROM sqlite_master WHERE name = 'installed'").await;
            assert!(tables.is_empty(), "the first statement was rolled back with the rest");

            execute_batch(&scratch.db(), &["CREATE TABLE installed (x INTEGER)".to_string()])
                .await
                .expect("a good batch runs");
            let tables = rows(&mut conn, "SELECT name FROM sqlite_master WHERE name = 'installed'").await;
            assert_eq!(tables, vec!["installed"]);
        });
    }

    #[test]
    fn a_missing_database_is_refused_not_created() {
        block_on(async {
            let scratch = Scratch::new();
            let err = revert(&scratch.db(), 0, None).await.expect_err("nothing to open");
            assert!(err.starts_with("Could not open the database"), "{err}");
            assert!(!scratch.db().exists(), "no empty database was created");
        });
    }
}
